ledger/ledger-consensual/src/main/kotlin/net/corda/v5/ledger/consensual/Party.kt

```diff
@@ -0,0 +1,16 @@
+package net.corda.v5.ledger.consensual
+
+import net.corda.v5.base.annotations.CordaSerializable
+import net.corda.v5.base.annotations.DoNotImplement
+import net.corda.v5.base.types.MemberX500Name
+import net.corda.v5.crypto.toStringShort
+import java.security.PublicKey
+
+@DoNotImplement
+@CordaSerializable
+interface Party {
```

Comment:
Do we want the party to be always a well known party? Will we need anonymous parties? @MrMatthewLayton mentioned account parties? 
What do you think @mnesbit ?

Reply:
I've always had a few concerns with regards to the `Party` implementation in Corda 4:

**Naming**
The root of the type hierarchy is `AbstractParty`. As far as I'm aware from looking at some Java APIs, the `Abstract` prefix seems to occur **only** when you have an interface for something, and an abstract of that thing, where there would be a naming clash; for example:

```
interface List<T> { ... }

abstract class AbstractList<T> : List<T> { ... }

class ArrayList<T> : AbstractList<T> { ... }
```

Therefore the type hierarchy would have been better described as:

```
abstract class Party
 > class AnonymousParty : Party
 > class WellKnownParty : Party
```

**Extensibility**
I'm not sure whether there was ever an intention to extend `AbstractParty` outside of the platform itself. It almost feels like it should have been declared as `sealed` rather than `abstract`, however, allowing external extensibility has allowed me to build an account model that hooks right into the `AbstractParty` implementation...

https://github.com/onix-labs/onixlabs-corda-identity-framework/blob/main/onixlabs-corda-identity-framework-contract/src/main/kotlin/io/onixlabs/corda/identityframework/contract/accounts/AccountParty.kt

I'm aware that in Corda 5, there is a heavy push towards using interfaces rather than using abstract or sealed classes, however this introduces a trade-off between things that _may_ leak abstraction, and things that are protected from external implementation, and IIRC, sealed interfaces are still experimental.

**Conclusion**
As @blsemo has pointed out, 
 - Do we want anonymous parties? 
 - Do we want accounts that can be implemented as parties?
 - Do we want the ability for consumers to implement these interfaces externally?

Reply:
We need a design meeting on this, because identity is a difficult topic. I certainly don't think that Party is fit for purpose.

Reply:
Let's go with this `Party` definition for now and revisit when I'm back. Everyone feel free to think about it...
I have raised https://r3-cev.atlassian.net/browse/CORE-5936 to tackle this in a few weeks time.

Reply:
I have moved Matthew's response here: https://r3-cev.atlassian.net/browse/CORE-5936